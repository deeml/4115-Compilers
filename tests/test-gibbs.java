import java.util.*;
import java.lang.*;

class program 
{
//input

//output

private static Random __rand = new Random(System.currentTimeMillis());
private static class bivar
{
public static double x = 0.0;
public static double y = 0.0;

public static double mean(double[] d) { double sum = 0.0;
for(double i : d) { sum += i; } return sum/((double) d.length);
}
public static double[] run()
{
double[] res = new double[2];
double[] __x = new double[5000];
x = __rand.nextFloat();
double[] __y = new double[5000];
y = __rand.nextFloat();
for (int i=0; i<5000; i++){
for (int j=0;j<100;j++){
 x = gamma_sample(3., y * y + 4., 0.);
y = random_sample(/*1. / x + 1, 1. / Math.sqrt(2 * x + 2)*/);
}
__x[i] = x;
__y[i] = y;
}
x = mean(__x);
y = mean(__y);
res[0] = x;
res[1] = y;
return res;
}
}private static double gamma_sample(double alpha, double beta, double lambda)
{
double gamma = 0.0;
gamma = 0.;
if (alpha <= 0. || beta <= 0.)
{
System.out.println("alpha and beta must be positive.");
return 0 - 1.;
}
if (alpha < 1.)
{
double b = 0.0;
double p = 0.0;
int flag = 0;
flag = 0;
b = 1 + alpha * 0.3678794;
while (flag == 0) {
p = b * __rand.nextFloat();
if (p > 1.)
{
gamma = 0 - Math.log(b - p / alpha);
if (__rand.nextFloat() <= Math.pow(gamma,alpha - 1.))
{
flag = 1;
}
}
else
{
gamma = Math.pow(p,1. / alpha);
if (__rand.nextFloat() <= Math.pow(2.718282,0 - gamma))
{
flag = 1;
}
}
}
}
else
{
if (alpha == 1.)
{
gamma = 0 - Math.log(__rand.nextFloat());
}
else
{
double y = 0.0;
y = 0 - Math.log(__rand.nextFloat());
while (__rand.nextFloat() > Math.pow(y * 1. - y,alpha - 1.)) {
y = 0 - Math.log(__rand.nextFloat());
}
gamma = alpha * y;
}
}
return beta * gamma + lambda;
}
private static double random_sample()
{
double u = 0.0;
u = __rand.nextFloat();
if (u < 0.5)
{
return Math.sqrt(0 - 0.5 * 3.141593 * Math.log(1 - Math.pow(2 * u - 1,2)));
}
else
{
return 0 - Math.sqrt(0 - 0.5 * 3.141593 * Math.log(1 - Math.pow(1 - 2 * u,2)));
}
}

public static void main(String[] args){
__rand = new Random(1234);
double[] res = bivar.run();
System.out.println(res[0]);
System.out.println(res[1]);

return ;
}
}
