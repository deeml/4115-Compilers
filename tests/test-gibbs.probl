input{}
output{}
fun random_sample(): float {
  float u;
  u = rand();
  if (u<0.5) {
    return sqrt(-0.5*3.141593*log(1-(2*u-1)^2));
  } else {
    return -sqrt(-0.5*3.141593*log(1-(1-2*u)^2));
  }
}

/* Return a random number drawn from a Gamma distribution with mean alpha*beta+lamba and variance alpha*beta^2. */
fun gamma_sample(float alpha, float beta, float lambda): float {
  float gamma;
  gamma=0.0;
  if (alpha <= 0.0 or beta <= 0.0) {
     print('alpha and beta must be positive.');
     return -1.0;
  }
  if (alpha < 1.0) {
    float b;
    float p;
    int flag;
    flag=0;
    b=1+alpha*0.3678794;
    while (flag == 0) {
      p=b*rand();
      if (p>1.0) {
        gamma=-log((b-p)/alpha);
        if (rand()<=gamma^(alpha-1.0)) {
          flag=1;
        }
      }
      else {
        gamma=p^(1.0/alpha);
        if (rand()<=2.718282^(-gamma)) { 
          flag=1;
        }
      }
    }
  } else {
    if (alpha == 1.0) {
      gamma = -log(rand());
    } else {
      float y;
      y = -log(rand());
      while (rand() > (y * (1.0 - y))^(alpha - 1.0)) {
        y = -log(rand());
      }
      gamma = alpha * y;
    }
  }
  return beta*gamma+lambda;
}

/* sample from p(x,y)=kx^2*exp(-xy^2-y^2+2y-4x) */
fun~ bivar(): x, y {
  x = gamma_sample(3.0,y*y+4.0,0.0);
  y = random_sample(1.0/(x+1),1.0/sqrt(2*x+2));
}

model
{
  srand(1234);
  x, y ~ bivar();
  print(x);
  print(y);
}
