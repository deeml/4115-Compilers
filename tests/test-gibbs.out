15.49813897084303
1.1034555555096368
